src/Std/Data/DHashMap/Raw.lean

```diff
@@ -0,0 +1,438 @@
+  b.buckets.foldlM (fun acc l => l.foldlM f acc) init
+
+/-- Folds the given function over the mappings in the hash map in some order. -/
+@[inline] def foldl (f : δ → (a : α) → β a → δ) (init : δ) (b : Raw α β) : δ :=
+  Id.run (b.foldlM f init)
+
+/-- Folds the given function over the mappings in the hash map in some order. -/
+@[inline] def forM (f : (a : α) → β a → m PUnit) (b : Raw α β) : m PUnit :=
+  b.buckets.forM (AssocList.forM f)
+
+/-- Support for the `for` loop construct in `do` blocks. -/
+@[inline] def forIn (f : (a : α) → β a → δ → m (ForInStep δ)) (init : δ) (b : Raw α β) : m δ :=
+  b.buckets.forIn init (fun bucket acc => bucket.forInStep acc f)
+
+instance : ForM m (Raw α β) ((a : α) × β a) where
+  forM m f := m.forM (fun a b => f ⟨a, b⟩)
+
+instance : ForIn m (Raw α β) ((a : α) × β a) where
+  forIn m init f := m.forIn (fun a b acc => f ⟨a, b⟩ acc) init
+
+/-- Transforms the hash map into a list of mappings in some order. -/
+@[inline] def toList (m : Raw α β) : List ((a : α) × β a) :=
+  m.foldl (fun acc k v => ⟨k, v⟩ :: acc) []
+
+/-- Transforms the hash map into an array of mappings in some order. -/
+@[inline] def toArray (m : Raw α β) : Array ((a : α) × β a) :=
+  m.foldl (fun acc k v => acc.push ⟨k, v⟩) #[]
+
+@[inline, inherit_doc Raw.toList] def Const.toList {β : Type v} (m : Raw α (fun _ => β)) :
+    List (α × β) :=
+  m.foldl (fun acc k v => ⟨k, v⟩ :: acc) []
+
+@[inline, inherit_doc Raw.toArray] def Const.toArray {β : Type v} (m : Raw α (fun _ => β)) :
+    Array (α × β) :=
+  m.foldl (fun acc k v => acc.push ⟨k, v⟩) #[]
+
+/-- Returns a list of all keys present in the hash map in some order. -/
+@[inline] def keys (m : Raw α β) : List α :=
+  m.foldl (fun acc k _ => k :: acc) []
+
+/-- Returns an array of all keys present in the hash map in some order. -/
+@[inline] def keysArray (m : Raw α β) : Array α :=
+  m.foldl (fun acc k _ => acc.push k) #[]
+
+/-- Returns a list of all values present in the hash map in some order. -/
+@[inline] def values {β : Type v} (m : Raw α (fun _ => β)) : List β :=
+  m.foldl (fun acc _ v => v :: acc) []
+
+/-- Inserts multiple mappings into the hash map by iterating over the given collection and calling
+`insert`. If the same key appears multiple times, the last occurrence takes precendence. -/
+@[inline] def insertMany [BEq α] [Hashable α] {ρ : Type w} [ForIn Id ρ ((a : α) × β a)]
+    (m : Raw α β) (l : ρ) : Raw α β :=
+  if h : 0 < m.buckets.size then
+    (Raw₀.insertMany ⟨m, h⟩ l).1
+  else m -- will never happen for well-formed inputs
+
+@[inline, inherit_doc Raw.insertMany] def Const.insertMany {β : Type v} [BEq α] [Hashable α]
+    {ρ : Type w} [ForIn Id ρ (α × β)] (m : Raw α (fun _ => β)) (l : ρ) : Raw α (fun _ => β) :=
+  if h : 0 < m.buckets.size then
+    (Raw₀.Const.insertMany ⟨m, h⟩ l).1
+  else m -- will never happen for well-formed inputs
+
+@[inline, inherit_doc Raw.insertMany] def Const.insertManyUnit [BEq α] [Hashable α] {ρ : Type w}
```

Comment:
I think this one doesn't have a docstring to inherit.

Is this a reasonable thing to have in a public API? When should a user prefer this over a `HashSet`?

Reply:
Yes, the only reason this exists is to be called by `HashSet.insertMany`. I have updated the docstring to say as much.

Reply:
LGTM!